codeworld-server/src/Util.hs

```diff
@@ -296,3 +299,53 @@ removeDirectoryIfExists dirName =
     handleExists e
         | isDoesNotExistError e = return ()
         | otherwise = throwIO e
+
+loadDumpedTree :: BuildMode -> UserId -> IO DirTree
+loadDumpedTree bm uid = do
+    let file = userProjectDir bm uid </> "tree.info"
+    exists <- doesFileExist file
+    realTree <- getDirectoryTree bm uid
+    case exists of
+        False -> return realTree
+        True -> do
+            content  <- LB.readFile file
+            return $ case decode content of
+                Nothing -> realTree
+                Just dumpedTree -> case compareTrees realTree (Dir "root" dumpedTree) of
+                    True  -> Dir "root" dumpedTree
+                    False -> realTree
+
+getDirectoryTree :: BuildMode -> UserId -> IO DirTree
+getDirectoryTree bm uid = do
+    ftr <- getDirectory $ userProjectDir bm uid
+    let (Node rootPath children) = toTree ftr
+    children' <- mapM (constructTree rootPath) $ filter notInfo children
+    return $ Dir "root" children'
+    where
+        constructTree :: FilePath -> Tree FilePath -> IO DirTree
+        constructTree prefix (Node path [(Node "dir.info" [])]) = do -- empty directory
+            name <- B.readFile $ prefix </> path </> "dir.info"
+            return $ Dir (T.decodeUtf8 name) []
+        constructTree prefix (Node path []) = do
```

Comment:
So it's only because of `dir.info` that this doesn't match an empty folder?  That seems like round-about logic.  A more direct way to decide would be to look at the filename: directories start with capital D, and projects start with S.

Reply:
`Node path [(Node "dir.info" [])]` match empty folder without projects. http://hackage.haskell.org/package/filesystem-trees-0.1.0.6/docs/System-File-Tree.html#t:FSTree does not paying attention is path file or directory. It is quckest way to determine this is empty directory on pattern matching level, not using any comparisons.